PR 1 of 2: Add RNA-seq Pathway Analysis - GSEA example (#405)
AlexsLemonade/refinebio-examples: 03-rnaseq/pathway-analysis_rnaseq_02_gsea.Rmd

```{r}
# Read in the contents of your differential expression results file
# `dge_url` can be replaced with a file path to a TSV file with your
# desired gene list results
dge_df <- readr::read_tsv(dge_url)
```

`read_tsv()` can read TSV files online and doesn't necessarily require you download the file first. 
Let's take a look at what the contrast results from the differential expression analysis looks like. 

```{r}
dge_df
```

## Getting familiar with `clusterProfiler`'s options

Let's take a look at what organisms the package supports.

```{r}
msigdbr_species()
```

MSigDB contains 8 different gene set collections [@Subramanian2005].

    H: hallmark gene sets
    C1: positional gene sets
    C2: curated gene sets
    C3: motif gene sets
    C4: computational gene sets
    C5: GO gene sets
    C6: oncogenic signatures
    C7: immunologic signatures

The data we're interested in here comes from human samples, so we can obtain just the gene sets (from all of the gene set collections listed above) relevant to _Homo sapiens_ by specifying `species = "Homo sapiens"`.

```{r}
hs_gene_sets <- msigdbr(

>   species = "Homo sapiens" # Replace with species name relevant to your data

Reviewer: I would not use all the gene sets here. It's not something I think we want to encourage. We'd rather folks select gene sets based on their question of interest. I do wonder if including _all_ of them might be contributing to the weirdness in #408. KEGG is pretty commonly used.

We can identify differentially regulated genes by filtering these results and use this list as input to GSEA.

Instead of using the URL below, you can use a file path to a TSV file with your desired gene list results.
First we will assign the URL to its own variable called, `dge_url`. 

```{r}
# Define the url to your differential expression results file
dge_url <- "https://refinebio-examples.s3.us-east-2.amazonaws.com/03-rnaseq/results/SRP078441/SRP078441_differential_expression_results.tsv"
```

Read in the file that has differential expression results. 
Here we are using the URL we set up above, but this can be a local file path instead _i.e._ you can replace `dge_url` in the code below with a path to file you have on your computer like: `file.path("results", "SRP078441_differential_expression_results.tsv")`.

```{r}
# Read in the contents of your differential expression results file
# `dge_url` can be replaced with a file path to a TSV file with your
# desired gene list results
dge_df <- readr::read_tsv(dge_url)
```

`read_tsv()` can read TSV files online and doesn't necessarily require you download the file first. 

> Let's take a look at what these contrast results from the differential expression analysis look like. 

Reviewer 2: Are there multiple contrasts here? I think it's just the one comparison right?

Author: Ah yes good point 👍

  species = "Homo sapiens" # Replace with species name relevant to your data
)
```

If you run the chunk above with specifying `category = "H"` to the `msigdbr()` function, it will return only the Hallmark gene sets for _Homo sapiens_.
See `?msigdbr` for more options.

Let's preview what's in `hs_gene_sets`.

```{r}
head(hs_gene_sets)
```

Looks like we have a data frame of gene sets with associated gene symbols and Entrez IDs.

In our differential expression results data frame, `dge_df` we have Ensembl gene identifiers. 
So we will need to convert our Ensembl IDs into either gene symbols or Entrez IDs for GSEA.

## Gene identifier conversion

We're going to convert our identifiers in `dge_df` to gene symbols, but you can, with the change of a single argument, use the same code to convert to many other types of identifiers!

The annotation package `org.Hs.eg.db` contains information for different identifiers [@Carlson2020-human].
`org.Hs.eg.db` is specific to _Homo sapiens_ -- this is what the `Hs` in the package name is referencing.

We can see what types of IDs are available to us in an annotation package with `keytypes()`.

```{r}
keytypes(org.Hs.eg.db)
```

> Even though we'll use this package to convert from Ensembl gene IDs (`ENSEMBL`) to gene symbols (`SYMBOL`), we could just as easily use it to convert from an Ensembl transcript ID (`ENSEMBLTRANS`) to Entrez IDs (`ENTREZID`).

Reviewer 2: Could make this comment more general since there's nothing in particular about transcripts that might be useful to the user. 
```suggestion
Even though we'll use this package to convert from Ensembl gene IDs (`ENSEMBL`) to gene symbols (`SYMBOL`), we could just as easily use it to convert to and from any of these `keytypes()` listed above. 
```

> Take a look at our other gene identifier conversion examples for examples with different species and gene ID types:

Reviewer 2: I think we can just have one spot that we refer people to the gene ID conversion examples and I think after we run the gene conversion example is better. So I'd just move this sentence and replace what's at 293. 

(I know this is copied over from how we have it written in microarray -- I don't feel strongly enough about this to make an issue, but it is something I want to switch if we do some edits too those notebooks later). So IDK.

Let's see a preview of `dge_mapped_df`.

```{r}
head(dge_mapped_df)
```

We will want to keep in mind that GSEA requires that genes are ranked from most highly positive to most highly negative and weighted according to their gene-level statistic, which is essential to the calculation of the enrichment score (ES), a pathway-level statistic.
GSEA also requires unique gene identifiers to produce the most accurate results.
This means if any duplicate gene identifiers are found in our dataset, we will want to retain the instance with the higher absolute value as this will be the instance most likely to be ranked most highly negative or most highly positive.
Otherwise the enrichment score results may be skewed and the `GSEA()` function will throw a warning.

Let's check to see if we have any gene symbols that mapped to multiple Ensembl IDs.

```{r}
any(duplicated(dge_mapped_df$gene_symbol))
```

Looks like we do have duplicated gene symbols.
Let's find out which gene symbols have been duplicated.

```{r}

> dup_gene_symbols <- dge_mapped_df %>%

Reviewer 2: We could probably trim out some of these previews to be a bit more succinct. But I don't feel strongly about this either.

Author: I agree, I took away the preview as we will see the duplicate gene symbols with their associated values in the next step.

```{r}
dge_mapped_df %>%
  dplyr::filter(gene_symbol %in% dup_gene_symbols) %>%
  dplyr::arrange(gene_symbol)
```

We can see that the associated values vary for each row.

As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSEA steps later, so let's keep the gene symbols associated with the higher absolute value of the log2 fold change.
GSEA relies on genes' rankings on the basis of this statistic.
Retaining the instance of the gene symbol with the higher absolute value means that we will retain the value that is likely to be more highly- or lowly-ranked or, put another way, the values less likely to be towards the middle of the ranked gene list.
We should keep this decision in mind when interpreting our results.
For example, if the duplicate identifiers tended to be enriched in a particular gene set, we may get an optimistic view of how perturbed that gene set is by preferentially selecting values that have a higher absolute value.

```{r}
filtered_dge_mapped_df <- dge_mapped_df %>%
  # Sort so that the highest absolute values of the log2 fold change are at the top
  dplyr::arrange(dplyr::desc(abs(log2FoldChange))) %>%

>   # Filter out the duplicated rows using `dplyr::distinct()`-- this will keep

Reviewer 2: I think this very long comment should be just a quick sentence before we do this action and then it can be shortened here. Some people don't read comments so this will make it more likely they don't miss it plus then we can avoid this multiline comment which makes this code chunk look bigger than it is.

  # the first row with the duplicated value thus keeping the row with the
  # highest absolute value of the log2 fold change
  dplyr::distinct(gene_symbol, .keep_all = TRUE)
```

Note that the log2 fold change values we use here have been transformed by `DESeq2` to account for genes with low counts or highly variable counts.
See the [`DESeq2` package vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) for more information on how `DESeq2` handles the log2 fold change values with the `lfcShrink()` function.

Let's check to see that we removed the duplicate gene symbols and kept the rows with the higher absolute value of the log2 fold change.

```{r}

> filtered_dge_mapped_df %>%

Reviewer 2: I think this could be shortened to just `any(duplicated())` type deal. We probably don't need to see this data previewed again.
